Replace data.frame(cbind()) with data.frame()

jdblischak opened the pull request:
I noticed the use of the pattern [`data.frame(cbind())`]( https://github.com/search?q=repo%3Akeaven%2FgsDesign%20data.frame(cbind(&type=code ) throughout the package, which seemed redundant to me. Surprisingly, this pattern is also common across [R packages on GitHub](https://github.com/search?utf8=%E2%9C%93&q=data.frame%28cbind+language%3AR&type=code). I've done various investigations to confirm that `data.frame(cbind())` and `data.frame()` are equivalent, at least for the uses in this package.

When combining vectors, the results are identical. When combining vectors and matrices, the results are identical unless the matrix has unnamed columns, in which case the auto-generated column names are slightly different. As far as I can tell, the columns are always named in this package.


<details>
<summary>Example code combining vectors and matrices
</summary>

```R
# Vectors only
vec1 <- 1
vec2 <- 1:2

# cbind() returns a matrix
cbind(vec1, vec2)
##      vec1 vec2
## [1,]    1    1
## [2,]    1    2
class(cbind(vec1, vec2))
## [1] "matrix" "array" 

# data.frame() returns a data frame
data.frame(vec1, vec2)
##   vec1 vec2
## 1    1    1
## 2    1    2
class(data.frame(vec1, vec2))
## [1] "data.frame"

# data.frame(cbind()) is equivalent to data.frame() when only using vectors
data.frame(cbind(vec1, vec2))
##   vec1 vec2
## 1    1    1
## 2    1    2
class(data.frame(cbind(vec1, vec2)))
## [1] "data.frame"

# Vectors and matrices
mat1 <- matrix(1:4, nrow = 2)
mat2 <- matrix(1:4, nrow = 2, dimnames = list(1:2, c("col1", "col2")))

# cbind() leaves missing column names as empty strings
cbind(vec1, vec2, mat1, mat2)
##   vec1 vec2     col1 col2
## 1    1    1 1 3    1    3
## 2    1    2 2 4    2    4
class(cbind(vec1, vec2, mat1, mat2))
## [1] "matrix" "array" 

# data.frame() converts missing column names to X#
data.frame(vec1, vec2, mat1, mat2)
##   vec1 vec2 X1 X2 col1 col2
## 1    1    1  1  3    1    3
## 2    1    2  2  4    2    4
class(data.frame(vec1, vec2, mat1, mat2))
## [1] "data.frame"

# data.frame(cbind()) converts missing column names to V#
data.frame(cbind(vec1, vec2, mat1, mat2))
##   vec1 vec2 V3 V4 col1 col2
## 1    1    1  1  3    1    3
## 2    1    2  2  4    2    4
class(data.frame(cbind(vec1, vec2, mat1, mat2)))
## [1] "data.frame"
```

</details>

Lastly, when using `data.frame(cbind())` when the first argument is already a data frame, this ends up calling `cbind.data.frame()` (via S3 dispatch), which is just a simple wrapper around `data.frame()` anyways, so it is essentially calling `data.frame(data.frame())`:

```R
body(cbind.data.frame)
## data.frame(..., check.names = FALSE)
```

I was also worried about the effect of `stringsAsFactors`. In R 4.0 the default was changed to `FALSE`, but {gsDesign} supports R >= 3.5, when it was still `TRUE`. However, `data.frame(cbind())` and `data.frame()` behave the same in regards to the default value of `stringsAsFactors`.

https://github.com/keaven/gsDesign/blob/6b9153c93b49334f84eac45945e6572c3097ad7a/DESCRIPTION#L19

<details>
<summary>Example code combining a data frame with additional vectors
</summary>

```R
R.version.string
## [1] "R version 3.5.0 (2018-04-23)"

df1 <- data.frame(col1 = letters[1:3], col2 = 1:3, col3 = c(1.1, 1.2, 1.3))
vec1 <- letters[4:6]
vec2 <- 4:6
cbind(df1, vec1, vec2)
##   col1 col2 col3 vec1 vec2
## 1    a    1  1.1    d    4
## 2    b    2  1.2    e    5
## 3    c    3  1.3    f    6
data.frame(df1, vec1, vec2)
##   col1 col2 col3 vec1 vec2
## 1    a    1  1.1    d    4
## 2    b    2  1.2    e    5
## 3    c    3  1.3    f    6
data.frame(cbind(df1, vec1, vec2))
##   col1 col2 col3 vec1 vec2
## 1    a    1  1.1    d    4
## 2    b    2  1.2    e    5
## 3    c    3  1.3    f    6
str(cbind(df1, vec1, vec2))
## 'data.frame':   3 obs. of  5 variables:
##  $ col1: Factor w/ 3 levels "a","b","c": 1 2 3
##  $ col2: int  1 2 3
##  $ col3: num  1.1 1.2 1.3
##  $ vec1: Factor w/ 3 levels "d","e","f": 1 2 3
##  $ vec2: int  4 5 6
str(data.frame(df1, vec1, vec2))
## 'data.frame':   3 obs. of  5 variables:
##  $ col1: Factor w/ 3 levels "a","b","c": 1 2 3
##  $ col2: int  1 2 3
##  $ col3: num  1.1 1.2 1.3
##  $ vec1: Factor w/ 3 levels "d","e","f": 1 2 3
##  $ vec2: int  4 5 6
str(data.frame(cbind(df1, vec1, vec2)))
## 'data.frame':   3 obs. of  5 variables:
##  $ col1: Factor w/ 3 levels "a","b","c": 1 2 3
##  $ col2: int  1 2 3
##  $ col3: num  1.1 1.2 1.3
##  $ vec1: Factor w/ 3 levels "d","e","f": 1 2 3
##  $ vec2: int  4 5 6
```

</details>

Additionally, I did some spot checks with `browser()` to convince myself the results were unchanged:

https://github.com/keaven/gsDesign/blob/6b9153c93b49334f84eac45945e6572c3097ad7a/R/gsMethods.R#L623
https://github.com/keaven/gsDesign/blob/6b9153c93b49334f84eac45945e6572c3097ad7a/R/gsMethods.R#L656
https://github.com/keaven/gsDesign/blob/6b9153c93b49334f84eac45945e6572c3097ad7a/R/gsSurv.R#L1386

Lastly, I checked [_Advanced R_](http://adv-r.had.co.nz/Data-structures.html#data-frames), and it explicitly advises against the pattern `data.frame(cbind())`:

> It's a common mistake to try and create a data frame by `cbind()`ing vectors together. This doesn't work because `cbind()` will create a matrix unless one of the arguments is already a data frame. Instead use `data.frame()` directly:
>
> The conversion rules for `cbind()` are complicated and best avoided by ensuring all inputs are of the same type.

nanxstats commented:
lgtm

jdblischak commented:
The new tests for `gsBoundSummary()` discovered a crucial difference between `data.frame()` and `data.frame(cbind())`: when one of the potential columns is `NULL`, `cbind()` discards it but `data.frame()` throws an error.

```R
── Failed tests ──────────────────────────────────────────────────────────────────────────────────────────
Error (test-independent-test-gsBoundSummary.R:179:3): Test CP, CP H1, and PP computations with multiple alpha values
Error in `data.frame(Efficacy, Futility, i = 1:(x$k - 1))`: arguments imply differing number of rows: 2, 0
Backtrace:
    ▆
 1. └─gsDesign::gsBoundSummary(...) at test-independent-test-gsBoundSummary.R:179:3
 2.   └─gsDesign:::gsBoundSummary0(...) at gsDesign/R/gsMethods.R:825:5
 3.     └─base::data.frame(Efficacy, Futility, i = 1:(x$k - 1)) at gsDesign/R/gsMethods.R:387:5

[ FAIL 1 | WARN 9 | SKIP 0 | PASS 1970 ]
```

What is the intention of the code (displayed below)? When `Futility` is `NULL`, is the goal to simply drop it from the output data frame? Or should it instead be `NA`? 

https://github.com/keaven/gsDesign/blob/7ad8beea26adcbb72bd80b85cf5fb0315c3de230/R/gsMethods.R#L385-L387

And a minimal code example:

```R
x <- 1:2
y <- NULL
z <- c(1.1, 1.2)

data.frame(cbind(x, y, z))
##   x   z
## 1 1 1.1
## 2 2 1.2
data.frame(x, y, z)
## Error in data.frame(x, y, z) : 
##   arguments imply differing number of rows: 2, 0

# Also note that `cbind()` only works properly here because the columns are all
# numeric. If there is character vector, everything is converted to character
# (because cbind() converts to a matrix, which requires only one data type)
w <- letters[1:2]
str(data.frame(cbind(x, y, z, w)))
## 'data.frame':	2 obs. of  3 variables:
##   $ x: chr  "1" "2"
## $ z: chr  "1.1" "1.2"
## $ w: chr  "a" "b"
```

jdblischak commented:
I confirmed that if we set the empty column to `NA_real_`, then `data.frame()` works fine. So the design question is: do we want to display the column `Futility` with missing values or remove it entirely from the output?

```R
x <- 1:2
y <- NA_real_
z <- c(1.1, 1.2)
data.frame(x, y, z)
##   x  y   z
## 1 1 NA 1.1
## 2 2 NA 1.2
str(data.frame(x, y, z))
## 'data.frame':	2 obs. of  3 variables:
##  $ x: int  1 2
##  $ y: num  NA NA
##  $ z: num  1.1 1.2
```

jdblischak commented:
I tried replacing `Futility <- NULL` with `Futility <- NA_real_`, but immediately got `rbind()` errors about mismatched number of columns, so I presume dropping the `Futility` column is the desired behavior.

@keaven could you please advise? Do you often use the strategy of setting a vector to `NULL` in order to drop it via `data.frame(cbind())`? Are there other motivations for using `data.frame(cbind())` that I haven't considered?

nanxstats commented:
Yeah, I think when Futility is `NULL`, the goal would be to simply drop it from the output data frame.

jdblischak commented:
> Yeah, I think when Futility is `NULL`, the goal would be to simply drop it from the output data frame.

Agreed. That's why I added back the `cbind()` for that particular example.

Looking at the other examples where I reduced it to only `data.frame()`, I think this should be fine. I didn't seen any obvious uses of assigning a column to `NULL` like in the `Futility <- NULL` example.

But it's hard for me to be 100% confident because not all the lines of code I modified here are covered by tests ([Codecov results for this PR](https://app.codecov.io/github/keaven/gsDesign/pull/182)). In general, it would be especially helpful to add some tests for the scenario where `outtype == 3`.

nanxstats commented:
Yep, that particular example looks like a special case where `data.frame(cbind())` becomes a expressive shortcut to handle the two possible code paths for column inclusion.

For most cases, a direct `data.frame()` call is canonical because it will fail properly if the above situation happens.

If something really breaks, hopefully we can easily bisect the Git history and track back to this PR and "unfix" things 🤞
